### ISSUE TITLE
BuildRequestBody failing with time.Time parameter

### ISSUE THREAD

**Issue by garbage-collector** [gophercloud/gophercloud#1233]
This code is failing:

```golang
func TestBuildRequestBody(t *testing.T) {
	type Params struct {
		ID   string    `json:"id"`
		Date time.Time `json:"date"`
	}

	params := &Params{
		ID:   "Foo",
		Date: time.Date(1970, time.January, 1, 0, 0, 0, 0, time.UTC),
	}

	body, err := BuildRequestBody(params, "params")
	th.AssertNoErr(t, err)

	th.AssertJSONEquals(t, `{ "params": { "id": "Foo", "date": "1970-01-01T00:00:00Z" } }`, body)
}
```

```
Failure in params_test.go, line 22: unexpected error "json: cannot unmarshal string into Go value of type map[string]interface {}"
```

The way ```BuildRequestBody``` is working tries to unmarshal a string into a map[string]interface{} because time.Time is considered as a struct I think. Looking for a fix, except if there is a workaround for this issue?

**Comment by garbage-collector** [comment]
After some digging, it looks like when the parameter type is a struct with a custom JSON marshaller that returns a ```string``` instead of a ```map[string]interface{}```, then the BuildRequestBody will never work.

**Comment by jtopjian** [comment]
Here's an example of how to handle `time.Time` within a request struct:

https://github.com/gophercloud/gophercloud/blob/master/openstack/keymanager/v1/orders/requests.go#L75

https://github.com/gophercloud/gophercloud/blob/master/openstack/keymanager/v1/orders/requests.go#L97-L110

**Comment by garbage-collector** [comment]
I just looked the ```keymanager/v1/orders``` package, and this is a way to handle ```time.Time```, but it looks more to a workaround than a solution.

And it looks more to a way of handling OpenStack time format than handling the issue in the BuildRequestBody code.

For example, we will still continue to have issues with this kind of code:

```
type ComplexField struct {
	ID   string
	Name string
}

func (c ComplexField) MarshalJSON() ([]byte, error) {
	return json.Marshal(c.ID + "." + c.Name)
}

func TestBuildRequestBody(t *testing.T) {
	type Params struct {
		ID      string       `json:"id"`
		Date    time.Time    `json:"date"`
		Complex ComplexField `json:"complex"`
	}

	params := &Params{
		ID:   "Foo",
		Date: time.Date(1970, time.January, 1, 0, 0, 0, 0, time.UTC),
		Complex: ComplexField{
			ID:   "Foo",
			Name: "Bar",
		},
	}

	body, err := BuildRequestBody(params, "params")
	th.AssertNoErr(t, err)

	th.AssertJSONEquals(t, `{"params": {"id": "Foo", "date": "1970-01-01T00:00:00Z", "complex": "Foo.Bar"}}`, body)
}
```

For now, ```time.Time``` is the only impacted struct of this bug, but going with a ```json:"-"``` tag does not allow to add ```required:"true"``` on fields that are struct, and it can end to other issues.

This is why I'm proposing a fix for this bug, as it helps to give to ```BuildRequestBody``` the expected behavior.

**Comment by jtopjian** [comment]
That's a fair assessment. On the other hand, though, the limitations you've pointed out have not caused any major issues to date. While current solutions can be thought of as workarounds, another way of looking at them is to make sure the various request structs are simple enough that they either work with `BuildRequestBody` as it currently is or with small modifications to the struct outside of it. 

I'm not opposed to your proposed change. I have two concerns about it:

* This might cause more focus to go into `BuildRequestBody` to support a variety of edge cases and lose focus on actually implementing various actions.
* Given I only have a handful of hours each week to look at Gophercloud, and since `BuildRequestBody` hasn't actually blocked anything major, I consider modifications to it a low priority of mine.

With that in mind, have you been blocked by this or is this more of an opportunity for an enhancement?

**Comment by garbage-collector** [comment]
I can completely agree on what you are saying.
TBH, I encountered this issue when working with a ```time.Time``` struct, but I fixed it because I needed a custom unmarshaller for time format purpose.

So no, this issue is definitely not blocking any PR for now. It was more an opportunity for an enhancement, but I can also agree that the issue appears in strange edge cases, so priority can be put on other issues than this one.

Feel free to close it if needed, I can also fully understand that it can create huger issue as this part of code is used everywhere in ```gophercloud``` package.

**Comment by jtopjian** [comment]
Let's leave this issue + the PR opened. Like I mentioned, I'm not opposed to the change, I just want to be cautious about it. Leaving this open will allow others to provide feedback, possibly with other use-cases that raise the priority. Plus it'll prevent duplicated work in this area.

As well, thank you for looking into this -- I'm in no way trying to negate the work you put into this!

**Comment by jtopjian** [comment]
I'm cleaning up some old issues and since this issue hasn't seen any input for over a year, I'm going to go ahead and close it. I'm happy to re-open it upon request, though.
